### PR TITLE
networkd: add scope to address section

### LINKED ISSUE #6446: networkd: scope for addresses
### Submission type

 - Request for enhancement (RFE)

### Description

Allow setting the scope of addresses configured on networks.

### Current Situation
```
[Network]
Addresses=169.254.10.1/30
```

currently results in

```
6: en0.110@en0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 00:1f:d0:20:a6:04 brd ff:ff:ff:ff:ff:ff
    inet 169.254.100.1/30 brd 169.254.100.3 scope global en0.110
```

### Problem
This makes it eligible for source address selection for routes via that interface, which is undesirable when the addresses are only used as a transfer network.

### Proposal
Add parameter `Scope` to `[Network]` with a value consistent with what iproute2 is offering

```
/sbin/ip addr add 169.254.100.1 dev en0.110 scope link
```
```
% ip addr add help 
[...]
SCOPE-ID := [ host | link | global | NUMBER ]
[...]
```

```
[Network]
Address=169.254.100.1/30
Scope=link
```

with result

```
6: en0.110@en0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 00:1f:d0:20:a6:04 brd ff:ff:ff:ff:ff:ff
    inet 169.254.100.1/30 brd 169.254.100.3 scope link en0.110
```

### PR DESCRIPTION
**Pull request by ssahani**
This work allows to configure address Scope to

host | link | global or a number.

Closes #6446

### CODE REVIEW

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -927,6 +927,52 @@ int config_parse_address_flags(const char *unit,
         return 0;
 }
 
+int config_parse_address_scope(const char *unit,
+                               const char *filename,
+                             unsigned line,
```
Comment (poettering) [inline src/network/networkd-address.c]: indentation is strange...

File: man/systemd.network.xml [hunk 2]
```diff
@@ -749,6 +749,14 @@
           </listitem>
         </varlistentry>
         <varlistentry>
+          <term><varname>Scope=</varname></term>
+          <listitem>
+            <para>The scope of the address, which can be <literal>global</literal>,
+            <literal>link</literal> or <literal>host</literal> or a unsigned integer. Defaults to
```
Comment (poettering) [inline man/systemd.network.xml]: "a unsigned" → "an unsigned"

Also, please document the valid range here. Is "0" permitted? and what is the largest value accepted?

File: src/network/networkd-address.c [hunk 3]
```diff
@@ -927,6 +927,52 @@ int config_parse_address_flags(const char *unit,
...
+                return r;
+
+        if (streq(rvalue, "host"))
+                n->scope = RT_SCOPE_HOST;
+        else if (streq(rvalue, "link"))
+                n->scope = RT_SCOPE_LINK;
+        else if (streq(rvalue, "global"))
+                n->scope = RT_SCOPE_UNIVERSE;
+        else {
+                r = safe_atou(rvalue, &k);
```
Comment (poettering) [inline src/network/networkd-address.c]: if scopes are uint32_t, we should parse this with safe_atou32() really!

**Comment by ssahani** [comment]
updated

File: src/network/networkd-address.c [hunk 4]
```diff
@@ -927,6 +927,49 @@ int config_parse_address_flags(const char *unit,
...
+                return r;
+
+        if (streq(rvalue, "host"))
+                n->scope = RT_SCOPE_HOST;
+        else if (streq(rvalue, "link"))
+                n->scope = RT_SCOPE_LINK;
+        else if (streq(rvalue, "global"))
+                n->scope = RT_SCOPE_UNIVERSE;
+        else {
+                r = safe_atou8(rvalue , &n->scope);
```
Comment (poettering) [inline src/network/networkd-address.c:961]: no space before the comma

**Comment by poettering** [comment]
CI failure appears unrelated

**Pull request merged**